### PR TITLE
Decouple some icky coupled bits in the Visual Studio project layer

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Our VS project layer has three highly coupled types: VisualStudioProjectTracker, DocumentProvider, and derivations of AbstractProject. Making these perform well and support background-thread calling is going to require rationalizing how data flows between these types, unlike the current model that is best described as "the world's largest ball of twine."

This PR is just cleaning up some other stuff in the way. DocumentProvider itself consumed a IVisualStudioHostProject interface which was implemented by AbstractProject and the Miscellaneous Files workspace's projects, which was getting it zero extra value. This meant that any refactoring of DocumentProvider would also impact MiscellanouesFilesWorkspace. The first commit here just deletes that link.

Once we did that, we can then delete IVisualStudioHostProject and IVisualStudioHostProjectContainer. This way the VisualStudioProjectTracker/DocumentProvider/AbstractProject coupling is more explicit rather than through an interface. If nothing else, it means finding dead code is easier.

<details><summary>Ask Mode template</summary>

### Customer scenario

This is code cleanup to help enable background initialization of the Visual Studio workspace.

### Risk

Moderate -- this code is fairly legacy and always has a chance of some subtle behavior creeping in.

### Performance impact

No change expected -- just changing some types and moving code around.

</details>

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This is exciting!

**Comment by CyrusNajmabadi** [comment]
Tagging @minestarks and @paulvanbrenk .  I think these types may be used by TS/JS.  I remember that i had to write some very coupled code here in the TS project system.  However, i think the change should def be possible.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs [hunk 1]
```diff
@@ -42,8 +41,11 @@ internal sealed partial class MiscellaneousFilesWorkspace : Workspace, IVsRunnin
         /// </summary>
         private IBidirectionalMap<uint, WorkspaceRegistration> _docCookieToWorkspaceRegistration = BidirectionalMap<uint, WorkspaceRegistration>.Empty;
 
-        private readonly Dictionary<ProjectId, HostProject> _hostProjects = new Dictionary<ProjectId, HostProject>();
-        private readonly Dictionary<uint, HostProject> _docCookiesToHostProject = new Dictionary<uint, HostProject>();
+        /// <summary>
+        /// The mapping of all doc cookies in the RDT and the <see cref="ProjectId"/> of the project and <see cref="SourceTextContainer"/> of the open
+        /// file we have created for that open buffer. An entry should only be in here if it's also already in <see cref="_docCookieToWorkspaceRegistration"/>.
+        /// </summary>
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs]: maybe document the threading requirements on this.

**Comment by CyrusNajmabadi** [comment]
I don't see anything wrong.  But i'm sure something will still end up breaking unexpectedly :)

**Comment by jasonmalinowski** [comment]
I didn't see uses in F#/TypeScript in //index of that interface but I should be more paranoid. I'm not sure how much of this is going to be possible to clean up without impacts there.

The other option for this whole piece of work is to write a new implementation of the CPS interfaces in parallel and then unify later. Trying to avoid "burn everything down" because we know how well that often works.

**Comment by CyrusNajmabadi** [comment]
Where's teh TS source again?  I'm happy to take a look since there are a couple of file that are important that are involved with project system integration.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi As far as I know, internal only.

**Comment by paulvanbrenk** [comment]
@CyrusNajmabadi yeah, internal only

**Comment by paulvanbrenk** [comment]
It looks like TypeScript should be fine here. The easiest way to check is to open a `.html` and/or `.aspx` file and make sure scriptblocks work as expected.

**Comment by jasonmalinowski** [comment]
Checking, neither TypeScript nor F# use any of the interfaces I'm deleting. I'm doing some extra testing now. @Pilchie OK for ask mode?

**Comment by jasonmalinowski** [comment]
F# testing passes (creating projects, some misc files transitions). Looks like TypeScript I broke though. Grr.

**Comment by Pilchie** [comment]
@jasonmalinowski Once it works I'll approve it :)

**Comment by jasonmalinowski** [comment]
@Pilchie Now with less TypeScript breakage!

**Comment by Pilchie** [comment]
Approved @jasonmalinowski - thanks!

**Pull request merged**